pkg/image/api/helper.go

```diff
@@ -313,12 +314,24 @@ func ManifestMatchesImage(image *Image, newManifest []byte) (bool, error) {
 	if err != nil {
 		return false, err
 	}
-	sm := schema1.SignedManifest{Raw: newManifest}
-	raw, err := sm.Payload()
-	if err != nil {
-		return false, err
+	var canonical []byte
+	if image.DockerImageManifestMediaType == schema2.MediaTypeManifest {
+		var m schema2.DeserializedManifest
+		if err := json.Unmarshal(newManifest, &m); err != nil {
+			return false, err
+		}
+		_, canonical, err = m.Payload()
+		if err != nil {
+			return false, err
+		}
+	} else {
```

Comment:
Yes, let's properly error on unknown manifest types.

On Thu, Jun 23, 2016 at 10:14 AM, Michal Minar <notifications@github.com>
wrote:

> In pkg/image/api/helper.go
> <https://github.com/openshift/origin/pull/8938#discussion_r68241768>:
>
> > @@ -313,12 +314,24 @@ func ManifestMatchesImage(image *Image, newManifest []byte) (bool, error) {
> >  	if err != nil {
> >  		return false, err
> >  	}
> > -	sm := schema1.SignedManifest{Raw: newManifest}
> > -	raw, err := sm.Payload()
> > -	if err != nil {
> > -		return false, err
> > +	var canonical []byte
> > +	if image.DockerImageManifestMediaType == schema2.MediaTypeManifest {
> > +		var m schema2.DeserializedManifest
> > +		if err := json.Unmarshal(newManifest, &m); err != nil {
> > +			return false, err
> > +		}
> > +		_, canonical, err = m.Payload()
> > +		if err != nil {
> > +			return false, err
> > +		}
> > +	} else {
>
> I still think we should handle a case of unknown manifest so we're not
> surprised when someone pushes schema v3.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/8938/files/084a33d8949b2810e9b1031593c9eb1df8d77407..a49051127bf0d42d650ec67f6b2b2d0077d3123d#r68241768>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABG_p_rE_Ctd1gRVrn0ItpqaEI8Mprxaks5qOpS4gaJpZM4IiLYZ>
> .
>